### ISSUE TITLE
link error : icu56.1, python.3.5.0, pyicu1.9.2  OSX Yosemite.

### ISSUE THREAD

**Issue by maveryKearney** [ovalhub/pyicu#8]
In the interest of getting a different perspective on the compile problem on Windows 10, I moved to OSX. I've had working the combination icu55.1, python 3.5.0, pyicu 1.9.2 on OSX Yosemite. However, when I advanced to icu56.1, 

- icu56.1 built and tested w/o error
- Using python 3.5.0 download for MacOSX (generic) from 
      https://www.python.org/downloads/release/python-350/
    - built pyicu 1.9.2 without error via 
         python3 setup.py build
         (i.e. I am trying to be as generic as I can)
    - installed without error via
        python3 setup.py install
   BUT
       python3 setup.py test fails

Actually, I encounter the following error, (same as in test) when I just attempt to import  icu:

hurst:PyICU-1.9.2 mkearney$ python3
Python 3.5.0 (v3.5.0:374f501f4567, Sep 12 2015, 11:00:19) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> import icu
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mkearney/bryllyg/opt/icu/56.1/PyICU-1.9.2/icu.py", line 37, in <module>
    from docs import *
  File "/Users/mkearney/bryllyg/opt/icu/56.1/PyICU-1.9.2/docs.py", line 23, in <module>
    from _icu import *
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so, 2): Library not loaded: libicui18n.56.dylib
  Referenced from: /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so
  Reason: image not found
>>> quit

I actually encountered this when I first tried integrating with my built from source python.3.5.0.
So the idea above was to eliminate everything custom at my end except the machine and os
I am running on. 

With luck, there is still something obviously wrong that I just am unaware of. It sure looks that way.

I am thinking I need to drop back to 55.1 which I did have working and is good enough for me, though
I have encountered 56.1 has fixed some bugs that I have worked around.

**Comment by ovalhub** [comment]
> On Oct 16, 2015, at 13:29, maveryKearney <notifications@github.com> wrote:
> 
> In the interest of getting a different perspective on the compile problem on Windows 10, I moved to OSX. I've had working the combination icu55.1, python 3.5.0, pyicu 1.9.2 on OSX Yosemite. However, when I advanced to icu56.1,
> 
> icu56.1 built and tested w/o error
> Using python 3.5.0 download for MacOSX (generic) from https://www.python.org/downloads/release/python-350/
> built pyicu 1.9.2 without error via python3 setup.py build (i.e. I am trying to be as generic as I can)
> installed without error via python3 setup.py install BUT python3 setup.py test fails
> Actually, I encounter the following error, (same as in test) when I just attempt to import icu:
> 
> hurst:PyICU-1.9.2 mkearney$ python3
> Python 3.5.0 (v3.5.0:374f501f4567, Sep 12 2015, 11:00:19) 
> [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> 
> import requests
> import icu
> Traceback (most recent call last):
> File "", line 1, in 
> File "/Users/mkearney/bryllyg/opt/icu/56.1/PyICU-1.9.2/icu.py", line 37, in 
> from docs import *
> File "/Users/mkearney/bryllyg/opt/icu/56.1/PyICU-1.9.2/docs.py", line 23, in 
> from _icu import *
> ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so, 2): Library not loaded: libicui18n.56.dylib
> Referenced from: /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so
> Reason: image not found
> quit
> 
Well, the errors tells you that the dynamic linker could not find libicui18n.56.dylib. Where do you have it installed ? In a standard place ? If not, you might have to muck with the DL_ env vars to help the dynamic linker. Also, use otool to inspect libicui18n.56.dylib and make sure you have everything it depends on.

Andi..

**Comment by maveryKearney** [comment]
That's exactly what I have been thinking, 
which is why I went for the standard package.

**Comment by maveryKearney** [comment]
I believe I have confirmed for my satisfaction that I have the problem with icu56.1 and not with icu55.1.
I am now only working with std python 3.5.0 MacOSx(64 bit) prebuilt release. I was making errors with DYLD_LIBRARY_PATH. 
Here is a log with traceback showing the error I am seeing now. Having installed ic56.1 in python 3.5.0 with pyicu1.9.1, I start python and import icu.  By the way, I install using pyicu's setup.py, since that is what I have had to do with source built python in the past. Regardless, I do not encounter this problem wtih python 3.5.0/ pyciu 1.9.2 and icu55.1 integration.

-------
hurst:PyICU-1.9.2 mkearney$ cd ~
hurst:~ mkearney$ /Library/Frameworks/Python.framework/Versions/3.5/bin/python3
Python 3.5.0 (v3.5.0:374f501f4567, Sep 12 2015, 11:00:19) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import icu
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/icu.py", line 37, in <module>
    from docs import *
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/docs.py", line 23, in <module>
    from _icu import *
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so, 2): Symbol not found: __ZNK6icu_5514LEFontInstance10canDisplayEi
  Referenced from: /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so
  Expected in: flat namespace
 in /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so

------------------------------

**Comment by ovalhub** [comment]
> On Oct 17, 2015, at 05:41, maveryKearney <notifications@github.com> wrote:
> 
> I believe I have confirmed for my satisfaction that I have the problem with icu56.1 and not with icu55.1.
> I am now only working with std python 3.5.0 MacOSx(64 bit) prebuilt release. I was making errors with DYLD_LIBRARY_PATH. 
> Here is a log with traceback showing the error I am seeing now. Having installed ic56.1 in python 3.5.0 with pyicu1.9.1, I start python and import icu. By the way, I install using pyicu's setup.py, since that is what I have had to do with source built python in the past. Regardless, I do not encounter this problem wtih python 3.5.0/ pyciu 1.9.2 and icu55.1 integration.
> 
> hurst:PyICU-1.9.2 mkearney$ cd ~
> hurst:~ mkearney$ /Library/Frameworks/Python.framework/Versions/3.5/bin/python3
> Python 3.5.0 (v3.5.0:374f501f4567, Sep 12 2015, 11:00:19) 
> [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> 
> import icu
> Traceback (most recent call last):
> File "", line 1, in 
> File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/icu.py", line 37, in 
> from docs import *
> File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/docs.py", line 23, in 
> from _icu import *
> ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so, 2): Symbol not found: __ZNK6icu_5514LEFontInstance10canDisplayEi
> Referenced from: /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so
> Expected in: flat namespace
> in /Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/PyICU-1.9.2-py3.5-macosx-10.6-intel.egg/_icu.cpython-35m-darwin.so
> 
Ah, that's different. This usually happens when you use one version of icu/pyicu during compilation (header files at least) and another at runtime.
In the version used at runtime the signature of LEFontInstance::canDisplay() changed or disappeared.

Andi..

**Comment by ovalhub** [comment]
Just for the record: I now built PyICU 1.9.2 with ICU 56.1 on Mac OS X 
10.10.3 with Python 2.7.5 and it builds and passes all tests just fine.

Trying the latest Python 3.5 now.

Andi..

**Comment by maveryKearney** [comment]
Andi,

 

Continuing the exploration of the Mac side of things.

 

I am still seeing the same import  error integrating the components above. I
duplicated my results of last week

starting from scratch with:

1.       python 350. Mac OS X  64-bit/32-bit installer.  To my surprise it
installs the 32-bit version on my machine by default.

Consequently there are ld warnings at the end of the PyICU build. I could
imagine that such a compilation error could cause the import error when
importing icu.

2.       From the python350 source tar-ball.   This builds by default a 64
bit version, I don't see ld warnings at the end of the PyIcu 1.0.2 build. 

 

However, I see the same import error message that I originally reported in
both cases.  It appears

that the ld warnings from the different platform difference is of no
consequence.

 

To the best of my knowledge I have no code in this collection. I have walked
through sys.path to make certain only the intended modules are on sys.path

 

My next step is to verify my other claim from the end of last week, that I
don't have the import error problem with  icu55.1 and pyicu 1.9.2.

 

If that works, I will return to the Win 10 and Win 7 machines to see I can
get icu 55.1 and pyicu 1.9.2 working

on those OS's. It would be sufficient for me to use icu 55 but I see that in
56.1, they have some bug fixes

for problems I have encountered, though I have worked around them. 

 

A complication in all this is that python from now on is using Visual studio
2015, but icu developers  use (I think) visual C++ 2010, though I get the
impression they are in the midst of upgrading to VS 2015.

 

That's where I stand at this moment.  I have a feeling you have been unable
to reproduce the error I am reporting.

 

Regards, 

-m

**Comment by maveryKearney** [comment]
Andi,

 

I just verified again that I don't have the import error problem with
icu55.1 and pyicu 1.9.2

I verified it with both the installation executable and source-built
versions of python 3.5.0.

 

So tomorrow, I'll go back to the windows machine.

 

Regards, 

-m

**Comment by ovalhub** [comment]
> On Oct 19, 2015, at 13:25, maveryKearney <notifications@github.com> wrote:
> 
> Andi,
> 
> Continuing the exploration of the Mac side of things.
> 
> I am still seeing the same import error integrating the components above. I
> duplicated my results of last week
> 
> starting from scratch with:
> 
> 1. python 350. Mac OS X 64-bit/32-bit installer. To my surprise it
> installs the 32-bit version on my machine by default.

Don't install the binaries, compile it from source  That way, you're sure you're using the same version of the tool chain for everything. It takes 10 minutes.

Also, use the latest xcode command line tools. I had an older version first and it dumped core compiling Python 3.5.0.

> Consequently there are ld warnings at the end of the PyICU build. I could
> imagine that such a compilation error could cause the import error when
> importing icu.
> 
> 2. From the python350 source tar-ball. This builds by default a 64
> bit version, I don't see ld warnings at the end of the PyIcu 1.0.2 build. 
> 
> 
> 
> However, I see the same import error message that I originally reported in
> both cases. It appears
> 
> that the ld warnings from the different platform difference is of no
> consequence.
> 
> 
> 
> To the best of my knowledge I have no code in this collection. I have walked
> through sys.path to make certain only the intended modules are on sys.path
> 
> 
> 
> My next step is to verify my other claim from the end of last week, that I
> don't have the import error problem with icu55.1 and pyicu 1.9.2.
> 
> 
> 
> If that works, I will return to the Win 10 and Win 7 machines to see I can
> get icu 55.1 and pyicu 1.9.2 working
> 
> on those OS's. It would be sufficient for me to use icu 55 but I see that in
> 56.1, they have some bug fixes
> 
> for problems I have encountered, though I have worked around them. 
> 
> 
> 
> A complication in all this is that python from now on is using Visual studio
> 2015, but icu developers use (I think) visual C++ 2010, though I get the
> impression they are in the midst of upgrading to VS 2015.
> 
> 
> 
> That's where I stand at this moment. I have a feeling you have been unable
> to reproduce the error I am reporting.

I have not tried this on Windows as I don't have access to a Windows system anymore.

I have built icu 56.1, Python 3.5.0, PyICU 1.9.2, all with the same compiler - the latest xcode 7 command line tools - on a the latest Mac OS x 10.10.
I've had no errors and had to set no DL_ env var either. All tests pass. I have not reproduced your problem.

Andi..

**Comment by maveryKearney** [comment]
Andi,

 

  Regarding  your comment,   “I've had no errors and had to set no DL_ env var either.”

I presume you mean having  to set DYLD_LIBRARY_PATH. Yes?  I definitely missed

that detail.  I’ll  have to look at that more closely.  I think that would explain

the problem I am seeing.  

 

I do not perform the install step when I build icu. If I understand correctly,

the libraries are thus not installed in /usr/local/lib. I want icu to be rock

solid stable with particular installation of my code. I experiment

with different icu data tables, and my application uses different 

tables for different languages. Also, I do not want icu data tables to 

change as a consequence of an accidental icu upgrade required by some other

software on a system. Perhaps there are better ways to accomplish these same

goals. Owning DL paths has been my means so far.

 

If icu 56 now only looks into /usr/local/lib for the libraries, in my case it will probably 

find an old edition of the libraries from a prior install. That would explain the 

link errors.

 

Unfortunately, I am on the road right now, and don’t have access to system

to test this.

 

Thanks for the feedback.

-m

**Comment by ovalhub** [comment]
> On Oct 21, 2015, at 00:39, maveryKearney <notifications@github.com> wrote:
> 
> Andi,
> 
> 
> 
> Regarding your comment, “I've had no errors and had to set no DL_ env var either.”
> 
> I presume you mean having to set DYLD_LIBRARY_PATH. Yes?

Yes, that's what I meant.

Andi..

> I definitely missed
> 
> that detail. I’ll have to look at that more closely. I think that would explain
> 
> the problem I am seeing. 
> 
> 
> 
> I do not perform the install step when I build icu. If I understand correctly,
> 
> the libraries are thus not installed in /usr/local/lib. I want icu to be rock
> 
> solid stable with particular installation of my code. I experiment
> 
> with different icu data tables, and my application uses different 
> 
> tables for different languages. Also, I do not want icu data tables to 
> 
> change as a consequence of an accidental icu upgrade required by some other
> 
> software on a system. Perhaps there are better ways to accomplish these same
> 
> goals. Owning DL paths has been my means so far.
> 
> 
> 
> If icu 56 now only looks into /usr/local/lib for the libraries, in my case it will probably 
> 
> find an old edition of the libraries from a prior install. That would explain the 
> 
> link errors.
> 
> 
> 
> Unfortunately, I am on the road right now, and don’t have access to system
> 
> to test this.
> 
> 
> 
> Thanks for the feedback.
> 
> -m
> 
> 
> 
> 
> 
> 
> 
> 
> 
> From: ovalhub [mailto:notifications@github.com] 
> Sent: Tuesday, October 20, 2015 3:40 PM
> To: ovalhub/pyicu <pyicu@noreply.github.com>
> Cc: maveryKearney <ma.kearney@verizon.net>
> Subject: Re: [pyicu] link error : icu56.1, python.3.5.0, pyicu1.9.2 OSX Yosemite. (#8)
> 
> 
> 
>

**Comment by ovalhub** [comment]
> On Oct 21, 2015, at 00:39, maveryKearney <notifications@github.com> wrote:
> 
> Andi,
> 
> 
> 
> Regarding your comment, “I've had no errors and had to set no DL_ env var either.”
> 
> I presume you mean having to set DYLD_LIBRARY_PATH. Yes? I definitely missed
> 
> that detail. I’ll have to look at that more closely. I think that would explain
> 
> the problem I am seeing. 
> 
> 
> 
> I do not perform the install step when I build icu. If I understand correctly,
> 
> the libraries are thus not installed in /usr/local/lib. I want icu to be rock
> 
> solid stable with particular installation of my code. I experiment
> 
> with different icu data tables, and my application uses different 
> 
> tables for different languages. Also, I do not want icu data tables to 
> 
> change as a consequence of an accidental icu upgrade required by some other
> 
> software on a system. Perhaps there are better ways to accomplish these same
> 
> goals. Owning DL paths has been my means so far.
> 
> 
> 
> If icu 56 now only looks into /usr/local/lib for the libraries, in my case it will probably 
> 
> find an old edition of the libraries from a prior install. That would explain the 
> 
> link errors.

When you build PyICU, you must also add the correct include and linker flags in setup.py so that the intended version of icu is picked.- if you have multiple versions installed and in non-standard locations.

Andi..

> 
> 
> 
> Unfortunately, I am on the road right now, and don’t have access to system
> 
> to test this.
> 
> 
> 
> Thanks for the feedback.
> 
> -m
> 
> 
> 
> 
> 
> 
> 
> 
> 
> From: ovalhub [mailto:notifications@github.com] 
> Sent: Tuesday, October 20, 2015 3:40 PM
> To: ovalhub/pyicu <pyicu@noreply.github.com>
> Cc: maveryKearney <ma.kearney@verizon.net>
> Subject: Re: [pyicu] link error : icu56.1, python.3.5.0, pyicu1.9.2 OSX Yosemite. (#8)
> 
> 
> 
>

**Comment by maveryKearney** [comment]
Andi,

      You seem to be correct regarding DYLD_LIBRARY_PATH.  I say seem only because I wish it were not so.

Discussions of DYLD_LIBRARY_PATH around the web suggest that DYLD_... is not present in El Capitan. Once I 

installed the icu lib which puts the dlls  in /usr/local/lib, everything worked. 

 

LD_LIBRARY_PATH has been around a long time in unix land and I was pleased to see it existed in OSX.

I was using it to support different icu data packages. I found I could change the library and change the icu

package that users of my app would see.   Perhaps there is a better way to achieve the same thing,

and I just stumbled into a hack.

 

       Regardless, I also just stumbled into a common development debug scenario. I am seeing different

behavior between icu 55.1 and 56.1. In the past I have just run two processes each pointing to different

library and then compare the behavior. Now I think I need two machines.  I only have one MAC at this

moment.  Of course I could run a windows machine as one of the two cases. Or just move development back to Windows

machine where I can still point to  multiple libraries

 

       So thanks for your input.  I still need to get back to the problem that started all this, i.e. 

the pyicu 1.9.2 compilation problem on windows which I hacked around. Perhaps it will go away

with icu 56.1

 

regards,

-m
